src/shelly_hap_temperature_sensor.cpp

```diff
@@ -0,0 +1,154 @@
+}
+
+std::string TemperatureSensor::name() const {
+  return cfg_->name;
+}
+
+Status TemperatureSensor::SetConfig(const std::string &config_json,
+                                    bool *restart_required) {
+  struct mgos_config_se cfg = *cfg_;
+  cfg.name = nullptr;
+  json_scanf(config_json.c_str(), config_json.size(), "{name: %Q, unit: %d",
+             &cfg.name, &cfg.unit);
+
+  mgos::ScopedCPtr name_owner((void *) cfg.name);
+  // Validation.
+  if (cfg.name != nullptr && strlen(cfg.name) > 64) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s",
+                        "name (too long, max 64)");
+  }
+  if (cfg.unit < 0 || cfg.unit > 1) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid unit");
+  }
+  // Now copy over.
+  if (cfg_->name != nullptr && strcmp(cfg_->name, cfg.name) != 0) {
+    mgos_conf_set_str(&cfg_->name, cfg.name);
+    *restart_required = true;
+  }
+  if (cfg_->unit != cfg.unit) {
+    cfg_->unit = cfg.unit;
+  }
+  return Status::OK();
+}
+
+Status TemperatureSensor::SetState(const std::string &state_json) {
+  (void) state_json;
+  return Status::OK();
+}
+
+void TemperatureSensor::ValueChanged() {
+  current_temperature_characteristic_->RaiseEvent();
+}
+
+Status TemperatureSensor::Init() {
+  uint16_t iid = svc_.iid + 1;
+  current_temperature_characteristic_ = new mgos::hap::FloatCharacteristic(
+      iid++, &kHAPCharacteristicType_CurrentTemperature, 0.0, 100.0, 0.1,
+      [this](HAPAccessoryServerRef *server UNUSED_ARG,
+             const HAPFloatCharacteristicReadRequest *request UNUSED_ARG,
+             float *value) {
+        auto tempval = temp_sensor_->GetTemperature();
+        if (!tempval.ok()) {
+          return kHAPError_Busy;
+        }
+        float temp = static_cast<float>(tempval.ValueOrDie());
+        *value = truncf(temp * 10) / 10;
+        return kHAPError_None;
+      },
+
+      true /* supports_notification */, nullptr /* write_handler */,
+      kHAPCharacteristicDebugDescription_CurrentTemperature);
+  AddChar(current_temperature_characteristic_);
+  AddChar(new mgos::hap::UInt8Characteristic(
+      iid++, &kHAPCharacteristicType_TemperatureDisplayUnits, 0, 255, 1,
+      [this](HAPAccessoryServerRef *server UNUSED_ARG,
+             const HAPUInt8CharacteristicReadRequest *request UNUSED_ARG,
+             uint8_t *value) {
+        *value = cfg_->unit;
+        return kHAPError_None;
+      },
+      true /* supports_notification */,
+      [this](HAPAccessoryServerRef *server UNUSED_ARG,
+             const HAPUInt8CharacteristicWriteRequest *request UNUSED_ARG,
+             uint8_t value) {
+        if (value >= 0 && value <= 1) {
+          cfg_->unit = value;
+        }
+        return kHAPError_None;
+      },
+      kHAPCharacteristicDebugDescription_TemperatureDisplayUnits));
+  temp_sensor_->notifier_ = std::bind(&TemperatureSensor::ValueChanged, this);
+  LOG(LL_INFO, ("Exporting Temp"));
+
+  temp_sensor_->StartUpdating(5000);
```

Comment:
this should be a config setting, exposed in a UI

Reply:
Done. I would restrict this to 1s minimum, as the conversion time is usually 750ms